blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py

```diff
@@ -104,54 +425,140 @@ def group_files_by_date(strings):
-    with db_config.managed_session() as session:
-        for upload_date, image_names in grouped_images_by_date.items():
-            upload_date = datetime.strptime(upload_date, "%Y-%m-%d_%H_%M_%S")
-            for image_name in image_names:
-                query = session.query(DetectionInformation).filter(
-                    DetectionInformation.image_customer_name == customer_name,
-                    DetectionInformation.image_upload_date == upload_date,
-                    DetectionInformation.image_filename == image_name,
-                )
-                results = query.all()
-
-                if results:
-                    extracted_data = [result.__dict__ for result in results]
-                    if upload_date in images_statistics:
-                        images_statistics[upload_date].extend(extracted_data)
-                    else:
-                        images_statistics[upload_date] = extracted_data
-
-    return images_statistics
+    image_counts = {}
+
+    try:
+        db_config = connect_to_database(database_parameters_json)
+        db_config.create_connection()
+    
+        with db_config.managed_session() as session:
+            for upload_date, image_names in grouped_images_by_date.items():
+                logger.info(f'Upload Date: {upload_date} \n')
+                upload_date = datetime.strptime(upload_date, "%Y-%m-%d_%H_%M_%S")
+                logger.info(f'Formatted Upload Date: {upload_date} \n')
+                for image_name in image_names:
+                    query = session.query(DetectionInformation).filter(
+                        DetectionInformation.image_customer_name == customer_name,
+                        DetectionInformation.image_upload_date == upload_date,
+                        DetectionInformation.image_filename == image_name,
+                        DetectionInformation.conf_score > conf_score_threshold
+                    )
+                    results = query.all()
+                    count = len(results)
+
+                    # Populating image_counts
+                    image_key = (customer_name, upload_date, image_name)
+                    image_counts[image_key] = count
+
+                    # Populating images_statistics with detailed information
+                    if results:
+                        extracted_data = [result.__dict__ for result in results]
+                        if upload_date in images_statistics:
+                            images_statistics[upload_date].extend(extracted_data)
+                        else:
+                            images_statistics[upload_date] = extracted_data
+                            
+    except SQLAlchemyError as e:
+        logger.error(f"Database operation failed: {e}")
+    except ValueError as e:
+        logger.error(f"Configuration error: {e}")
+    except Exception as e:
+        logger.error(f"An unexpected error occurred: {e}")
+
+    return images_statistics, image_counts
```

Comment:
The image counts can be counted from image statistics no?
What if instead of using all these multiples arrays with information that makes the code hard to follow we just use a dataframe with the actual data and we just share that across the functions?